opslevel/resource_opslevel_domain.go

```diff
@@ -1,9 +1,224 @@
+	var data DomainResourceModel
+
+	// Read Terraform plan data into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	resource, err := r.client.CreateDomain(opslevel.DomainInput{
+		Description: opslevel.RefOf(data.Description.ValueString()),
+		Name:        opslevel.RefOf(data.Name.ValueString()),
+		Note:        opslevel.RefOf(data.Note.ValueString()),
+		OwnerId:     opslevel.NewID(data.Owner.ValueString()),
+	})
+	if err != nil {
+		resp.Diagnostics.AddError("opslevel client error", fmt.Sprintf("Unable to create domain, got error: %s", err))
+		return
+	}
+	data.Aliases = types.ListNull(types.StringType)
+	data.Id = types.StringValue(string(resource.Id))
+	data.LastUpdated = types.StringValue(timeLastUpdated())
+
+	// Write logs using the tflog package
+	// Documentation: https://terraform.io/plugin/log
+	tflog.Trace(ctx, "created a domain resource")
+
+	// Save data into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+}
+
+func (r *DomainResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	var data DomainResourceModel
+
+	// Read Terraform prior state data into the model
+	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	resource, err := r.client.GetDomain(data.Id.ValueString())
+	if err != nil {
+		resp.Diagnostics.AddError("opslevel client error", fmt.Sprintf("Unable to read domain, got error: %s", err))
+		return
+	}
+	domainAliases, d := types.ListValueFrom(ctx, types.StringType, resource.ManagedAliases)
+	resp.Diagnostics.Append(d...)
+
+	data.Aliases = domainAliases
+	data.Description = types.StringValue(resource.Description)
+	data.Name = types.StringValue(string(resource.Name))
+	data.Note = types.StringValue(resource.Note)
+	data.Owner = types.StringValue(string(resource.Owner.Id()))
+
+	// Save updated data into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+}
+
+func (r *DomainResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	var data DomainResourceModel
+
+	// Read Terraform plan data into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	resource, err := r.client.UpdateDomain(data.Id.ValueString(), opslevel.DomainInput{
+		Description: opslevel.RefOf(data.Description.ValueString()),
+		Name:        opslevel.RefOf(data.Name.ValueString()),
+		Note:        opslevel.RefOf(data.Note.ValueString()),
+		OwnerId:     opslevel.NewID(data.Owner.ValueString()),
+	})
+	if err != nil {
+		resp.Diagnostics.AddError("opslevel client error", fmt.Sprintf("Unable to update domain, got error: %s", err))
+		return
+	}
+	domainAliases, d := types.ListValueFrom(ctx, types.StringType, resource.ManagedAliases)
+	resp.Diagnostics.Append(d...)
+
+	data.Aliases = domainAliases
```

Comment:
starting to see how we could setup some generic functions for these kinds of common operations we'll need to do in ever resource.